kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+                    ) && last == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                    Option::None => old(self).index@ + 1 == ghost_state@.cells.len(),
+                },
+        {
+            let cur = self.next(ghost_state);
+            if cur.is_none() {
+                proof {
+                    assert(match last {
+                        Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                            (self.index@ - 1) as nat,
+                        ) && last == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                        Option::None => old(self).index@ + 1 == ghost_state@.cells.len(),
+                    });
+                }
+                break ;
+            } else {
+                proof {
+                    assert(match cur {
+                        Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                            (self.index@ - 1) as nat,
+                        ) && cur == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                        Option::None => false,
+                    });
+                }
+                last = cur;
+            }
+        }
+        last
+    }
+}
+
+impl<'a, T: ?Sized + ListNodeV<'a, T>> ListV<'a, T> {
+    pub closed spec fn well_formed_list(&self, ghost_state: &Tracked<GhostState<'a, T>>) -> bool {
+        &&& ghost_state@.cells.len() >= 1
+        &&& forall|i: nat| 0 <= i < ghost_state@.cells.len() ==> well_formed_node(ghost_state, i)
+        &&& forall|i: nat|
+            0 <= i < (ghost_state@.cells.len() - 1) as nat
+                ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                Option::Some(_) => true,
+                Option::None => false,
+            }
+        &&& match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+            Option::Some(_) => false,
+            Option::None => true,
+        }
+        &&& ghost_state@.cells[0].id() == self.head.0.id()
+    }
+
+    pub const fn new() -> (res: (ListV<'a, T>, Tracked<GhostState<'a, T>>))
+        ensures
+            res.1@.cells.len() == 1,
+            well_formed_node(&res.1, 0),
+            res.0.well_formed_list(&res.1),
+    {
+        let (list_link, Tracked(points_to)) = ListLinkV::<T>::empty();
+        let cell = &list_link.0;
+        let mut ghost_state = Tracked(
+            GhostState { cells: Seq::empty(), points_to_map: Map::tracked_empty() },
+        );
+
+        proof {
+            ghost_state.borrow_mut().cells = ghost_state@.cells.push(cell);
+            ghost_state.borrow_mut().points_to_map.tracked_insert(0, points_to);
+        }
+
+        let list = ListV { head: list_link };
+
+        (list, ghost_state)
+    }
+
+    pub fn head(&self, ghost_state: &Tracked<GhostState<'a, T>>) -> (res: Option<&'a T>)
+        requires
+            ghost_state@.cells.len() >= 1,
```

Comment:
why don't we require the list to be `well_formed`? that would prevent restating some of these preconditions

Reply:
Because I do not want to manually trigger `forall` with a proof block of assert. I think it is sufficient for `head`, but you may choose an alternative way to fully prove `well_formed_list`. It should be possible to use `well_formed_node` on index 0, though.